### PR TITLE
Implement improved numerically-stable softplus

### LINKED ISSUE #260: Issue with softplus approximation
The current implemenation of softplus (aka log1pexp) uses two bounds to decide whether to return `0, log1p(exp(x)), or x`

https://github.com/pymc-devs/Theano-PyMC/blob/511c778f8a595444e009bcad738d552413b16f2c/theano/tensor/nnet/sigm.py#L356-L366

So far so good (almost). However, different bounds are used for the `c_code` depending on the precision being used https://github.com/pymc-devs/Theano-PyMC/blob/511c778f8a595444e009bcad738d552413b16f2c/theano/tensor/nnet/sigm.py#L376-L400

Unless I am missing something there might be an issue with both of the upper bounds used in the `c_code` (`14.0f and 16.0`). 

According to [Machler, 2012](https://cran.r-project.org/web/packages/Rmpfr/vignettes/log1mexp-note.pdf) who describes a refined method for `log1pexp` there is unnecessary error in using the approximation `log1pexp = x` for values of x < 33.3 (green line in Figure 3).

image

I don't know why the upper bounds are so low. It seems like `log1p(exp(x))` shouldn't overflow so soon. Using the code mentioned in the comment I get overflow only at +89 for float32 and at +710 for float64. This makes me very confident that the reason for choosing these was computational speed, but according to Machler (2012) a safer cutoff would have been 33.3, or an equally good intermediate approximation `x + exp(-x)` could be used until 33.3.

***

For reference here is the entire Machler (2012) suggestion, which tries to achieve a good compromise between speed and numerical precision for the entire real range:

```python
    if x <= -37.0: 
        return exp(x)
    elif x <= 18.0:
        return log1p(exp(x))
    elif x <= 33.3:
        return x + exp(-x)
    else:
        return x   
```
Plus the current appropriate lower bounds in `c_code` to return 0 instead of underflowing when x is very negative.

The changes are:
1. Using the approximation `exp(x)` for values below -37, instead of the current approach of returning `0` for values below 30 in `Theano` or the more costly `log1p(exp(x))` for any values that don't underflow in `c`.
2. Using the approximation `x + exp(-x)` in the range [18, 33.3] instead of the more costly (according to Machler and, it seems, the Theano guys as well) `log1p(exp(x))`
3. Using the approximation `x` only when it is no worse than `log1p(exp(x))` or `x + exp(-x)`, which occurs at 33.3 (again according to Machler).

### PR DESCRIPTION
**Pull request by ricardoV94**
This PR alters the Softplus Op to use the numerically stable approximation described by [Machler 2012](https://cran.r-project.org/web/packages/Rmpfr/vignettes/log1mexp-note.pdf).

It is also computationally efficient in that it uses safe and fast altenatives to the good default `log1p(exp(x))`:

1. `exp(x)` in the range `[-inf, -37]` (not used previously)
2. `log1p(exp(x))` in the range `[-37, 18]` (previously used in the range `[-30, 30]` in Python or `[-inf/underflow, 14/16]` in `c_code`)
2. `x + np.exp(-x)` in the range `[18, 33.3]` (not used previously)
3. `x` in the range `[33.3, +inf]` (previously used in the range `[30, +inf]` in Python or `[14/16, +inf]` in `c_code`)

The previous 64-bit `c_code` cutoff at `16.0` for when to use the approximation `log1p(x) = x` was a bit too eager (according to Machler, and my unittests below) and the rational for it was not clear either (the op was described as having the goal of numerical stabilization and not speedup). In addition, the Python version was not doing the same thing as the C counterpart.

According to my tests, there was no issue with the (also unexplained) 32-bit cutoff of `14.0`.

When it comes to unittests, the only thing that was checking for precision seems to have been the `TestSoftplus` in `test.sigm.py` via `verify_grad` which is not sensitive enough to detect the reduction in error introduced by this PR. It can, however, detect when very wrong cutoffs or obviously wrong approximations are used in the `c_code` so it is a good sanity check that the current PR is not breaking anything.

**Edit: Added specific accuracy unittests, which would fail with the previous 64-bit `c_code` implementation**

The 32bit wouldn't fail in the previous implementation due to its limited range, but I think it is better to keep the same implementation for simplicity sake. It is a bit more conservative/inefficient than needed, but not less accurate!

***

Here are a few important guidelines and requirements to check before your PR can be merged:
+ [x] There is an informative high-level description of the changes.
+ [x] The description and/or commit message(s) references the relevant GitHub issue(s).
+ [x] [`pre-commit`](https://pre-commit.com/#installation) is installed and [set up](https://pre-commit.com/#3-install-the-git-hook-scripts).
+ [x] The commit messages follow [these guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
+ [x] The commits correspond to [_relevant logical changes_](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes), and there are **no commits that fix changes introduced by other commits in the same branch/BR**.  If your commit description starts with "Fix...", then you're probably making this mistake.
+ [x] There are tests covering the changes introduced in the PR.

Closes #260

### CODE REVIEW

File: tests/tensor/nnet/test_sigm.py [hunk 1]
```diff
@@ -110,6 +110,7 @@ def setup_method(self):
 
     def test_elemwise(self):
         utt.verify_grad(softplus, [np.random.rand(3, 4)])
+        utt.verify_grad(softplus, [np.linspace(-40, 40, 20)])
```
Comment (brandonwillard) [inline tests/tensor/nnet/test_sigm.py:113]: In this repository, we're moving away from performing wasteful, time consuming, shotgun-like numerical tests, so we need to combine these `verify_grad`s into a single one that explicitly tests only important and/or necessary values.

**Review by brandonwillard — CHANGES_REQUESTED** [review]
Ideally, we wouldn't need to maintain our own numerical implementations (and their tests), instead we would only dispatch to established numerical libraries like NumPy and SciPy.  This module has already undergone related changes recently and I'm starting to think we should be looking to our "dispatch" libraries for a set of standard implementations (Python ones, at least) and stick with those.

Is there a NumPy/SciPy implementation that satisfies these accuracy concerns?  Also, note that there is a distinct `Op` for an "ultra fast" sigmoid; we could/should do the same for softplus, if we want to introduce alternative formulations for speed.

**Comment by twiecki** [comment]
> Is there a NumPy/SciPy implementation that satisfies these accuracy concerns?

Certainly not if we want a C Op implementation. Not sure we can avoid that.

**Comment by ricardoV94** [comment]
> Ideally, we wouldn't need to maintain our own numerical implementations (and their tests), instead we would only dispatch to established numerical libraries like NumPy and SciPy.  This module has already undergone related changes recently and I'm starting to think we should be looking to our "dispatch" libraries for a set of standard implementations (Python ones, at least) and stick with those.

> Is there a NumPy/SciPy implementation that satisfies these accuracy concerns?  

My understanding is that like for the `log1mexp`, there are no specialized implementations in numpy or scipy at the moment.

> Also, note that there is a distinct `Op` for an "ultra fast" sigmoid; we could/should do the same for softplus, if we want to introduce alternative formulations for speed.

I think those ultrafast methods provide a trade-off with worse precision. In this case the changed function should be faster but no less precise than using the default `log1p(exp(x))` for the entire real range. Since there is no trade-off in precision it doesn't seem to be worth it to separate it into different ops in this case (for this particular algorithm).

**Comment by brandonwillard** [comment]
> My understanding is that like for the `log1mexp`, there are no specialized implementations in numpy or scipy at the moment.

Yeah, I'm not aware of one for that, but we should first consider use functions like `scipy.special.[logsumexp, softmax]` for their respective `Op`s.

> The `log1mexp` does not exist as an op right now, and is instead implemented in the pymc math module. Do you see a reason to prefer one of the approaches above the other?

I think the current PyMC3 `log1mexp` approach is fine.

We would only want to create a new `Op` to wrap a new SciPy/NumPy function, or to use a very specific C/JAX implementation; otherwise, it's generally better to combine existing "primitive" `Op`s, especially since it allows for more optimizations.  Collapsing multiple `Op`s into a single `Op` can remove the chance to apply some optimizations (e.g. a single rewrite for `x / x -> 1` cannot be applied to an `Op` that represents `lambda x, y: 2 * x / y` when the argument `y = x` is given).

Now, there is one exception to this: we can create specialized `Op`s for compositions of simpler `Op`s and apply those _after_ the "optimizations" in a later "specialization" pass over a graph.  In this case, we probably don't want to advertise direct use of such `Op`s.  Also, these `Op`s obviously need to be accompanied by graph rewrites that apply them; otherwise, they wouldn't be used.



> Since there is no trade-off in precision it doesn't seem to be worth it to separate it into different ops in this case (for this particular algorithm).

Those `Op`s are another type of "specialization".  They can be applied explicitly by a user who wants the performance/precision trade-offs they provide; either through direct use of those `Op`s, or through a choice of broad performance-based optimization settings that automatically apply them (similar to the "specialization" stage).

**Comment by brandonwillard** [comment]
Anyway, I think these changes are probably worth merging, but my primary concern is that we don't unnecessarily add more time to our test suite.  Find a few important points in the domain that confirm your implementation works and provide full coverage; otherwise, we'll have to trust the underlying numerical libraries and the referenced paper for the rest.

**Comment by ricardoV94** [comment]
> Anyway, I think these changes are probably worth merging, but my primary concern is that we don't unnecessarily add more time to our test suite. Find a few important points in the domain that confirm your implementation works and provide full coverage; otherwise, we'll have to trust the underlying numerical libraries and the referenced paper for the rest.

I added what I hope is a lightweight enough test for the critical points of the new algorithm. The precision is set high enough that it would trigger with the previous implementation (64-bit), but not with the current one. 

Given the concern about reducing the speed of the tests, why not cut down the previous test based on the gradient from `[np.random.rand(3, 4)]` to `[np.random.rand(2, 3)]`? Not sure if much time would be saved, but if the test with 12 points has been working for years, maybe it is safe to go with 6 points now.

**Review by twiecki — APPROVED** [review]

File: theano/tensor/nnet/sigm.py [hunk 2]
```diff
@@ -348,22 +348,35 @@ def values_eq_approx_remove_low_prec(a, b):
...
-    This helps numerical stability.
+    r"""
+    Compute log(1 + exp(x)), also known as softplus or log1pexp
+
+    This function is numerically more stable than the naive approach.
+
+    For details, see
+    https://cran.r-project.org/web/packages/Rmpfr/vignettes/log1mexp-note.pdf
+
+    References
```
Comment (michaelosthege) [inline theano/tensor/nnet/sigm.py:359]: Indentation in this block is off

**Review by michaelosthege — CHANGES_REQUESTED** [review]
Greta work! (one nitpick)
What are the implications for the `grad` of this `COp`? Does it suffer from the same instabilities?

File: theano/tensor/nnet/sigm.py (hunk 2 above)
Comment (ricardoV94) [reply theano/tensor/nnet/sigm.py:359]: Thanks, fixed

**Comment by ricardoV94** [comment]
> What are the implications for the `grad` of this `COp`? Does it suffer from the same instabilities?

To be honest I have no idea. Looking at the grad, it seems to rely solely on the `Sigmoid Op`, but other Ops grad may rely on this one.

**Review by twiecki — APPROVED** [review]

**Comment by twiecki** [comment]
Thanks @ricardoV94!

**Pull request merged**